src/vm/types.rs

```diff
@@ -639,12 +646,36 @@ impl TypeSignature {
                     .ok_or_else(|| (a.clone(), b.clone()))
             }
 
-        if a.admits_type(&b) {
-            Ok(a)
-        } else if b.admits_type(&a) {
-            Ok(b)
-        } else {
-            Err((a,b))
+        // same goes for the option type
+        // this little monster is an attempt to avoid an unneccessary clone
+        //   I'm not sure there's a better way to do this. I think _maybe_
+        //     a match statement would work, but I think that'd be nasty too.
```

Comment:
I think this is about as good as it can get -- the borrow checker won't let you move `a` or `b` via a `return` within the `if let` branch.  If you want, you can either (1) add comments to explain what `Some(0)` and `Some(1)` mean, or (2) create a private enum of the various possibilities and use those instead.  Either one would clarify what's going on here.

Reply:
I added some comments